src/ucs/config/parser.c

```diff
@@ -379,6 +379,145 @@ int ucs_config_sprintf_time(char *buf, size_t max, void *src, const void *arg)
     return 1;
 }
 
+int ucs_config_sscanf_speed(const char *buf, void *dest, const void *arg)
+{
+    size_t *dst     = (size_t*)dest;
+    char    str[16] = {0};
+    int     offset  = 0;
+    size_t  divider;
+    size_t  units;
+    double  value;
+    int     num_fields;
+
+    num_fields = sscanf(buf, "%lf%16s", &value, str);
+    if (num_fields < 2) {
+        return 0; /* bytes or bits? */
+    }
+
+    ucs_assert(num_fields == 2);
+
+    switch (str[offset]) {
+    case 'T':
+        units = UCS_TBYTE;
+        break;
+    case 'G':
+        units = UCS_GBYTE;
+        break;
+    case 'M':
+        units = UCS_MBYTE;
+        break;
+    case 'K':
+        units = UCS_KBYTE;
+        break;
+    case 'B':
+    case 'b':
+        units = 1;
+        goto bytes;
+    default:
+        return 0;
+    }
+
+    offset++;
+
+bytes:
+    switch (str[offset]) {
+    case 'B':
+        divider = 1;
+        break;
+    case 'b':
+        divider = 8;
+        break;
+    default:
+        return 0;
+    }
+
+    offset++;
+
+    switch (str[offset]) {
+    case 'p':
+    case '/':
+        break;
+    case 's':
+        goto sec;
+    case 0:
+        goto done;
+    default:
+        return 0;
+    }
+
+    offset++;
+
+sec:
+    switch (str[offset]) {
+    case 's':
+        if (!str[offset + 1]) { /* end-of-line */
+            break;
+        }
+        /* no break here */
+    default:
+        return 0;
+    }
+
+done:
+    ucs_assert((divider == 1) || (divider == 8)); /* bytes or bits */
+    *dst = (size_t)(value * units / divider);
+
+    return 1;
+}
+
+int ucs_config_sprintf_speed(char *buf, size_t max, void *src, const void *arg)
+{
+    static const char * suffixes[] = {"", "K", "M", "G", "T", NULL};
+
+    const char **suffix = &suffixes[0];
+    size_t       value  = *(size_t*)src;
+
+    while ((value >= 1024) && ((value % 1024) == 0) && *(suffix + 1)) {
+        value /= 1024;
+        ++suffix;
+    }
+    snprintf(buf, max, "%zu%sBps", value, *suffix);
+    return 1;
+}
+
+int ucs_config_sscanf_ib_pci_speed(const char *buf, void *dest, const void *arg)
+{
+    ucs_config_ib_pci_speed_t *dst = (ucs_config_ib_pci_speed_t*)dest;
+    char                      *delim;
+
+    delim = strchr(buf, ':');
+    if (!delim) {
+        return 0;
+    }
+
+    if (!ucs_config_sscanf_speed(delim + 1, &dst->speed, arg)) {
+        return 0;
+    }
+
+    strncpy(dst->name, buf, ucs_min(delim - buf, sizeof(dst->name)));
+    /* force close string by zero */
+    dst->name[ucs_min(delim - buf, sizeof(dst->name) - 1)] = 0;
+
+    return 1;
+}
+
+int ucs_config_sprintf_ib_pci_speed(char *buf, size_t max, void *src, const void *arg)
+{
+    ucs_config_ib_pci_speed_t *speed = (ucs_config_ib_pci_speed_t*)src;
+    int                        len;
+
+    len = snprintf(buf, ucs_min(max, sizeof(speed->name)), "%s:", speed->name);
```

Comment:
pls check what snprintf returns exactly... it's not he length of the resulting string